components/infra-proxy-service/server/migrations.go

```diff
@@ -0,0 +1,119 @@
+package server
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os"
+	"path"
+
+	uuid "github.com/chef/automate/lib/uuid4"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/chef/automate/api/interservice/infra_proxy/migrations/response"
+	"github.com/chef/automate/api/interservice/infra_proxy/migrations/service"
+)
+
+// Takes up file name from service.MigrationDataService_UploadFileServer.MigrationId and creates the file in the same directory
+func saveFile(migrationId string, filename string, fileData bytes.Buffer) error {
+	err := os.Mkdir(migrationId, 0777)
+	if err != nil {
+		log.WithError(err).Error("Unable to create directory for migration id", migrationId)
+		return err
+	}
+	filePath := path.Join(migrationId, filename)
+	file, err := os.Create(filePath)
+	if err != nil {
+		log.WithError(err).Error("Unable to create file")
+		return err
+	}
+	_, err = fileData.WriteTo(file)
+	if err != nil {
+		return err
+	}
+	return nil
+
+}
+
+// UploadFile Takes the stream of data to upload a file
+func (s *Server) UploadFile(stream service.MigrationDataService_UploadFileServer) error {
+	req, err := stream.Recv()
+	serverId := req.ServerId
+	fileName := req.GetMeta().GetName()
+	ctx := context.Background()
+	migrationId, err := createMigrationId()
+	if err != nil {
+		log.WithError(err).Error("Unable to create migration id")
+		res := createResponseWithErrors(err, migrationId)
+		stream.SendAndClose(res)
+		return err
+	}
+
+	_, err = s.service.Migration.StartMigration(ctx, migrationId, serverId)
+	fileData := bytes.Buffer{}
+	s.service.Migration.StartFileUpload(ctx, migrationId, serverId)
+	for {
+		req, err := stream.Recv()
+
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
```

Comment:
Since we have started the migration at L52, any error while uploading should set the status of the migration to FailedState with an understandable error message. There should be two phases that needs to be updated:
1. FailedtoUpload
2. FailedtoMigrate